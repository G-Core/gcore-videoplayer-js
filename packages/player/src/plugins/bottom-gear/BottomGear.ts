import { UICorePlugin, template, Events as ClapprEvents, $ } from '@clappr/core'
import { trace } from '@gcorevideo/utils'
import assert from 'assert'

import { CLAPPR_VERSION } from '../../build.js'

import pluginHtml from '../../../assets/bottom-gear/bottomgear.ejs'
import '../../../assets/bottom-gear/gear.scss'
import '../../../assets/bottom-gear/gear-sub-menu.scss'
import gearIcon from '../../../assets/icons/new/gear.svg'
import gearHdIcon from '../../../assets/icons/new/gear-hd.svg'
import { ZeptoResult } from '../../types.js'
import { MediaControlEvents } from '../media-control/MediaControl'

const VERSION = '2.19.12'

const T = 'plugins.bottom_gear'

export enum GearEvents {
  RENDERED = 'rendered',
}

/**
 * An element inside the gear menu
 * @beta
 * @deprecated
 */
export type GearOptionsItem = 'quality' | 'rate' | 'nerd'

/**
 * @deprecated Use {@link GearOptionsItem} instead
 */
export type GearItemElement = GearOptionsItem

// TODO disabled if no items added

/**
 * `PLUGIN` that adds the gear button with an extra options menu on the right side of the {@link MediaControl | media control} UI
 * @beta
 * @remarks
 * The plugins provides a base for attaching custom settings UI in the gear menu
 *
 * Depends on:
 *
 * - {@link MediaControl}
 *
 * @example
 * You can use bottom gear to add custom settings UI to the gear menu.
 *
 * ```ts
 * import { BottomGear } from '@gcorevideo/player/plugins/bottom-gear';
 *
 * class CustomOptionsPlugin extends UICorePlugin {
 *   // ...
 *
 *   override get events() {
 *     return {
 *       'click #my-button': 'doMyAction',
 *     }
 *   }
 *
 *   private doMyAction() {
 *     // ...
 *   }
 *
 *   override render() {
 *     const bottomGear = this.core.getPlugin('bottom_gear');
 *     if (!bottomGear) {
 *       return this;
 *     }
 *     this.$el.html('<button class="custom-option">Custom option</button>');
 *     // Put rendered element into the gear menu
 *     bottomGear.addItem('custom').html(this.$el)
 *     return this;
 *   }
 *
 *   // alternatively, add an option with a submenu
 *   override render() {
 *     this.$el.html(template(templateHtml)({
 *       // ...
 *     })));
 *     return this;
 *   }
 *
 *   private addGearOption() {
 *     this.core.getPlugin('bottom_gear')
 *       .addItem('custom', this.$el)
 *       .html($('<button class="custom-option">Custom option</button>'))
 *   }
 *
 *   override bindEvents() {
 *     this.listenToOnce(this.core, ClapprEvents.CORE_READY, () => {
 *       const bottomGear = this.core.getPlugin('bottom_gear');
 *       assert(bottomGear, 'bottom_gear plugin is required');
 *       // simple case
 *       this.listenTo(bottomGear, GearEvents.RENDERED, this.render);
 *       // or with a submenu
 *       this.listenTo(bottomGear, GearEvents.RENDERED, this.addGearOption);
 *      });
 *   }
 * }
 * ```
 */
export class BottomGear extends UICorePlugin {
  private isHd = false

  /**
   * @internal
   */
  get name() {
    return 'bottom_gear'
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  /**
   * @internal
   */
  static get version() {
    return VERSION
  }

  private static readonly template = template(pluginHtml)

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'media-control-gear',
    }
  }

  /**
   * @internal
   */
  override get events() {
    return {
      'click #gear-button': 'toggleGearMenu',
    }
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenToOnce(this.core, ClapprEvents.CORE_READY, this.onCoreReady)
    this.listenTo(
      this.core,
      ClapprEvents.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onActiveContainerChanged,
    )
  }

  addItem(name: string, $subMenu?: ZeptoResult): ZeptoResult {
    const $existingItem = this.$el.find(`#gear-options li[data-${name}`)
    if ($existingItem.length) {
      trace(`${T} addItem already exists`, { name })
      return $existingItem
    }
    const $item = $('<li></li>')
      .attr(`data-${name}`, '')
      .appendTo(this.$el.find('#gear-options'))
    if ($subMenu) {
      trace(`${T} addItem adding submenu`, { name })
      $subMenu
        .addClass('gear-sub-menu-wrapper')
        .hide()
        .appendTo(this.$el.find('#gear-options-wrapper'))
      $item.on('click', (e: MouseEvent) => {
        trace(`${T} addItem submenu clicked`, { name })
        e.stopPropagation()
        $subMenu.show()
        this.$el.find('#gear-options').hide()
      })
    }
    return $item
  }

  private onActiveContainerChanged() {
    trace(`${T} onActiveContainerChanged`)
    this.bindContainerEvents()
  }

  private bindContainerEvents() {
    trace(`${T} bindContainerEvents`)
    this.listenTo(
      this.core.activeContainer,
      ClapprEvents.CONTAINER_HIGHDEFINITIONUPDATE,
      this.highDefinitionUpdate,
    )
  }

  private highDefinitionUpdate(isHd: boolean) {
    trace(`${T} highDefinitionUpdate`, { isHd })
    this.isHd = isHd
    this.$el.find('.gear-icon').html(isHd ? gearHdIcon : gearIcon)
  }

  /**
   * @internal
   */
  override render() {
    trace(`${T} render`)
    const mediaControl = this.core.getPlugin('media_control')
    if (!mediaControl) {
      return this // TODO test
    }
    const icon = this.isHd ? gearHdIcon : gearIcon
    this.$el
      .html(BottomGear.template({ icon }))
      .find('#gear-sub-menu-wrapper')
      .hide()

    // TODO make non-clickable when there are no items
    mediaControl.putElement('gear', this.$el)

    setTimeout(() => {
      this.trigger(GearEvents.RENDERED)
    }, 0)

    return this
  }

  /**
   * Collapses any submenu open back to the gear menu
   */
  refresh() {
    this.$el.find('.gear-sub-menu-wrapper').hide()
    this.$el.find('#gear-options').show()
  }

  private toggleGearMenu() {
    this.$el.find('#gear-options-wrapper').toggle()
  }

  private hide() {
    this.$el.find('#gear-options-wrapper').hide()
  }

  private onCoreReady() {
    trace(`${T} onCoreReady`)
    const mediaControl = this.core.getPlugin('media_control')
    assert(mediaControl, 'media_control plugin is required')
    this.listenTo(
      mediaControl,
      ClapprEvents.MEDIACONTROL_RENDERED,
      this.onMediaControlRendered,
    )
    this.listenTo(mediaControl, ClapprEvents.MEDIACONTROL_HIDE, this.hide)
  }

  private onMediaControlRendered() {
    trace(`${T} onMediaControlRendered`)
    this.render()
  }
}
