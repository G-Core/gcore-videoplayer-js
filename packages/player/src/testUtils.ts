import { $, Playback, UICorePlugin } from '@clappr/core'
import Events from 'eventemitter3'
import { vi } from 'vitest'
/**
 * @internal
 * @deprecated
 * TODO use createMockPlayback() instead
 */
export class _MockPlayback extends Events {
  constructor(
    protected options: any,
    readonly i18n: any,
    protected playerError?: any,
  ) {
    super()
  }

  get name() {
    return 'mock'
  }

  consent() {}

  play() {}

  pause() {}

  stop() {}

  destroy() {}

  seek() {}

  seekPercentage() {}

  getDuration() {
    return 100
  }

  enterPiP() {}

  exitPiP() {}

  getPlaybackType() {
    return Playback.LIVE
  }

  getStartTimeOffset() {
    return 0
  }

  getCurrentTime() {
    return 0
  }

  isHighDefinitionInUse() {
    return false
  }

  mute() {}

  unmute() {}

  volume() {}

  configure() {}

  attemptAutoPlay() {
    return true
  }

  canAutoPlay() {
    return true
  }

  onResize() {
    return true
  }

  trigger(event: string, ...args: any[]) {
    this.emit(event, ...args)
  }
}

export function createMockCore(
  options: Record<string, unknown> = {},
  container: any = createMockContainer(),
) {
  const el = document.createElement('div')
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    $el: $(el),
    activePlayback: container.playback,
    activeContainer: container,
    i18n: {
      t: vi.fn().mockImplementation((key: string) => key),
    },
    options: {
      ...options,
    },
    configure: vi.fn(),
    getPlaybackType: vi.fn().mockReturnValue(Playback.LIVE),
    getPlugin: vi.fn(),
    load: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockPlugin() {
  return Object.assign(new Events(), {
    enable: vi.fn(),
    disable: vi.fn(),
  })
}

export function createSpinnerPlugin() {
  return Object.assign(createMockPlugin(), {
    show: vi.fn(),
    hide: vi.fn(),
  })
}

export function createMockPlayback(name = 'mock') {
  const emitter = new Events()
  return Object.assign(emitter, {
    name,
    currentLevel: -1,
    el: document.createElement('video'),
    dvrEnabled: false,
    dvrInUse: false,
    levels: [],
    consent() {},
    play() {},
    pause() {},
    stop() {},
    destroy: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    getDuration: vi.fn().mockImplementation(() => 100),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getPlaybackType: vi.fn().mockImplementation(() => Playback.LIVE),
    getStartTimeOffset: vi.fn().mockImplementation(() => 0),
    getCurrentTime: vi.fn().mockImplementation(() => 0),
    isHighDefinitionInUse: vi.fn().mockImplementation(() => false),
    mute: vi.fn(),
    unmute: vi.fn(),
    volume: vi.fn(),
    configure: vi.fn(),
    attemptAutoPlay: vi.fn().mockImplementation(() => true),
    canAutoPlay: vi.fn().mockImplementation(() => true),
    onResize: vi.fn().mockImplementation(() => true),
    setPlaybackRate: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockContainer(playback: any = createMockPlayback()) {
  const el = playback.el
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    playback,
    $el: $(el),
    getDuration: vi.fn().mockReturnValue(0),
    getPlugin: vi.fn(),
    getPlaybackType: vi.fn().mockReturnValue(Playback.LIVE),
    isDvrInUse: vi.fn().mockReturnValue(false),
    isDvrEnabled: vi.fn().mockReturnValue(false),
    isPlaying: vi.fn().mockReturnValue(false),
    play: vi.fn(),
    seek: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockMediaControl(core: any) {
  const mediaControl = new UICorePlugin(core)
  mediaControl.$el.html(
    `<div class="media-control-left-panel" data-media-control></div>
    <div class="media-control-right-panel" data-media-control></div>
    <div class="media-control-center-panel" data-media-control></div>`,
  )
  const elements = {
    gear: $(document.createElement('div')),
  }
  // @ts-ignore
  mediaControl.getElement = vi.fn().mockImplementation((name) => elements[name])
  // @ts-ignore
  mediaControl.putElement = vi.fn()
  // @ts-ignore
  mediaControl.toggleElement = vi.fn()
  return mediaControl
}

export function createMockBottomGear(core: any) {
  const plugin: any = new UICorePlugin(core)
  plugin.getItem = vi.fn()
  plugin.addItem = vi.fn().mockImplementation((name: string, $el: any) => {
    const existing = plugin.$el.find(`[data-${name}]`)
    if (existing.length) {
      return existing
    }
    return $('<li></li>').attr(`data-${name}`, '').append($el).appendTo(plugin.$el)
  })
  plugin.refresh = vi.fn()
  return plugin
}
