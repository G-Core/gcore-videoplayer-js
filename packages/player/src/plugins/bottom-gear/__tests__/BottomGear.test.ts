import { MockedFunction, beforeEach, describe, expect, it, vi } from 'vitest'

import { BottomGear, GearEvents } from '../BottomGear'
import { createMockCore, createMockMediaControl } from '../../../testUtils'

describe('BottomGear', () => {
  let mediaControl: any
  let core: any
  let bottomGear: BottomGear
  let onGearRendered: MockedFunction<() => void>
  beforeEach(() => {
    core = createMockCore()
    mediaControl = createMockMediaControl(core)
    core.getPlugin = vi
      .fn()
      .mockImplementation((name) =>
        name === 'media_control' ? mediaControl : null,
      )
    bottomGear = new BottomGear(core)
    onGearRendered = vi.fn()
    bottomGear.on(GearEvents.RENDERED, onGearRendered, null)
    bottomGear.render()
  })
  it('should render', () => {
    expect(bottomGear.el.innerHTML).toMatchSnapshot()
  })
  it('should attach to media control', () => {
    expect(mediaControl.putElement).toHaveBeenCalledWith('gear', bottomGear.$el)
  })
  it('should emit event in the next cycle', async () => {
    expect(onGearRendered).not.toHaveBeenCalled()
    await new Promise((resolve) => setTimeout(resolve, 0))
    expect(onGearRendered).toHaveBeenCalled()
  })
  it('should render the gear menu hidden', () => {
    expect(bottomGear.$el.find('#gear-options-wrapper').css('display')).toBe(
      'none',
    )
  })
  describe('when clicked', () => {
    beforeEach(() => {
      bottomGear.$el.find('#gear-button').click()
    })
    it('should toggle the gear menu', () => {
      expect(bottomGear.$el.find('#gear-options-wrapper').css('display')).toBe(
        'block',
      )
    })
  })
})
