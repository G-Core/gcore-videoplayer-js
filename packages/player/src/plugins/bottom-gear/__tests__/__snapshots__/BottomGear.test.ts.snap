// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BottomGear > should render 1`] = `
"<button type="button" class="media-control-button gplayer-lite-btn gcore-skin-button-color gear-icon" id="gear-button">
    /assets/icons/new/gear.svg
</button>
<div class="gear-wrapper gcore-skin-bg-color" id="gear-options-wrapper" style="display: none;">
    <ul class="gear-options-list" id="gear-options"></ul>
</div>
"
`;
